libs/ml/src/ops/max_pool_1d.cpp

```diff
@@ -0,0 +1,200 @@
+  }
+}
+
+/**
+ * Computes gradient of 1D max pooling of kernel_size_ for each channel described here:
+ * http://ais.uni-bonn.de/papers/icann2010_maxpool.pdf
+ * Error signal of max pool is passed only to max node
+ * @param inputs vector of tensor references where at:
+ * inputs[0] = input_data[input_channels x input_height]
+ * @param error_signal tensor of size [output_channels=input_channels x
+ * number_of_stride_sized_steps]
+ * @return: output vector of tensors with back propagated error signal
+ * output[0]=input_error[inputs[0].shape]
+ */
+template <typename TensorType>
+std::vector<TensorType> MaxPool1D<TensorType>::Backward(const VecTensorType &inputs,
+                                                        const TensorType &   error_signal)
+{
+  assert(inputs.size() == 1);
+  assert(error_signal.shape() == ComputeOutputShape(inputs));
+
+  TensorType return_signal{inputs.at(0)->shape()};
+
+  auto output_shape = error_signal.shape();
+
+  SizeType iter;
+  DataType max;
+  DataType val;
+  SizeType max_iter;
+  auto     er_it = error_signal.cbegin();
+
+  for (SizeType n_i{0}; n_i < output_shape.at(2); n_i++)  // iterate over batch
+  {
+    for (SizeType i{0}; i < output_shape.at(1); i++)  // Iterate over kernel stride
+    {
+      iter = i * stride_size_;
+      for (SizeType c{0}; c < output_shape.at(0); ++c)  // Iterate over output channels
+      {
+        max      = inputs.at(0)->At(c, iter, n_i);
+        max_iter = iter;
+
+        // Find max node
+        for (SizeType j{0}; j < kernel_size_; j++)  // Iterate over kernel width
+        {
+          val = inputs.at(0)->At(c, iter + j, n_i);
+          if (val > max)
+          {
+            max      = val;
+            max_iter = iter + j;
+          }
+        }
+
+        // Add error to max node
+        return_signal.Set(c, max_iter, n_i, return_signal.At(c, max_iter, n_i) + *er_it);
+        ++er_it;
+      }
+    }
+  }
+
+  return {return_signal};
+}
+
+template <typename T>
+std::vector<fetch::math::SizeType> MaxPool1D<T>::ComputeOutputShape(
+    const VecTensorType &inputs) const
+{
+  std::vector<SizeType> output_shape;
+
+  // output_shape_[0]=number of output channels
+  output_shape.emplace_back(inputs.at(0)->shape().at(0));
+  // output_shape_[1]=number of stride_size steps over input size
+  output_shape.emplace_back((inputs.at(0)->shape().at(1) - (kernel_size_ - stride_size_)) /
+                            stride_size_);
+  // output_shape_[2]=batch dimension
+  output_shape.emplace_back(inputs.at(0)->shape().at(2));
+  return output_shape;
+}
+
+///////////////////////////////
+/// EXPLICIT INSTANTIATIONS ///
+///////////////////////////////
+
+// template class MaxPool1D<math::Tensor<int8_t>>;
+// template class MaxPool1D<math::Tensor<int16_t>>;
+template class MaxPool1D<math::Tensor<int32_t>>;
+template class MaxPool1D<math::Tensor<int64_t>>;
+// template class MaxPool1D<math::Tensor<uint8_t>>;
+// template class MaxPool1D<math::Tensor<uint16_t>>;
```

Comment:
These were commented because of an error (implicit conversion loses integer precision)

Reply:
This was resolved by adding a static cast to tensor.cpp file